lib/Index/Index.cpp

```diff
@@ -420,6 +423,29 @@ class ContainerTracker {
 
     return true;
   }
+
+  StackEntry identifyContainers(const TypeAliasDecl *D) const {
+    StackEntry Entry;
+    Entry.TrackedDecl = D;
+    Entry.ActiveKey = D;
+    if (auto TR = D->getUnderlyingTypeRepr()) {
+      associateTypeRepr(TR, D, Entry);
+    }
+    return Entry;
+  }
+
+  void associateTypeRepr(const TypeRepr *TR, const TypeAliasDecl *D,
+                         StackEntry &Entry) const {
+    if (auto composition = dyn_cast<CompositionTypeRepr>(TR)) {
```

Comment:
> Why does indexing look at TypeRepr and not the Type?

For the Contains relationship, it seemed like working into the existing mechanism using `ContainerTracker` would make sense. The container tracking, to emit the Contains relationship, works for `TypeRepr` but not `Type`.

https://github.com/swiftlang/swift/blob/8feb49724b41778d3f317d9226c99aec3ec51327/lib/Index/Index.cpp#L802-L805

I'm open to adding the relationship in a different way. Or using a different relationship, which would require handling it separately from the `ContainerTracker`. Trying to capture every permutation of the `TypeRepr` and associating them with the `typealias` Decl is definitely tricky, and might break in the future with new types.
 
What do you think is best here?

> I think this check will at the very least miss TypeReprs like (Int), as well as referenced nested inside functions and metatypes, like (Int) -> ()

😅 Good point! Those don't work. I'll add a test to cover those cases when we work out the other details.

Reply:
I guess the underlying issue is the disconnect between the indexstore representation and the problem we're trying to solve. If we consider this example:

```
typealias Foo = Bar
```

we'd like to have some kind of relationship to tell us that `Foo` is an alias of `Bar`, because right now we just see the reference to `Bar` but it's no related to `Foo` in any way. Likewise, if we had multiple types on the right hand side of the alias, we'd want to be able to distinguish the base type from other types, like generic arguments. In the following case:

```
typealias Foo = Bar<Baz>
```

we'd like to have the relationship to distinguish `Bar` (the thing we care about) from `Baz`.

But this starts to fall apart with non-nominal `TypeRepr`s, because what would we get from the following?

```
typealias Foo = [Bar]
typealias Foo = (Bar) -> Baz
```

In the first case, `Foo` is an alias for a specialization of `Array`, which doesn't appear in that `TypeRepr` at all. And the second case is even worse, because `Foo` is an alias for a function type that doesn't even have an equivalent nominal type to point back to, and just knowing that `Bar` and `Baz` were contained by `Foo` would tell us _something_, but it's not really what we want to know here. In any case, we'd have to combine the index data with syntax data to understand how the references are used by the alias.

It might just be that the indexstore isn't the right representation for what we're trying to track for typealiases, and it's the aliased type's USR that we really want (which we'd have to get from the AST some other way, like my hopeful [JSON AST PR](https://github.com/swiftlang/swift/pull/77321)), because then we could demangle it and determine precisely what it represents.

Reply:
The non-nominal `TypeRepr` seems like it really breaks this idea. I don't believe there is a good way to represent the information about "typealias of what?" for those types. I'm going to close this and reconsider. Thanks for the feedback folks!